Source/Genetic.py

```diff
@@ -122,9 +120,46 @@ def __PrintGrid(self, gridToPrint):
         for i in range(0, self.gridSize):
             print(gridToPrint[i])
 
+    #Find the valid paths between numbers and return a count of how many are valid
+    def __CountValidPaths(self, individual):
+        connectedNumbers = 0
+
+        for i in range(1, self.NumberofNumbers + 1):
+            isValid = True
+            #cursed lol, returns a list of all coordinates 9as tuples) of the given numer i
+            curNumCoordList = list(zip(np.where(np.array(individual) == i)[0], np.where(np.array(individual) == i)[1]))
+            for j in range(0, len(curNumCoordList) - 1):
+                currentX, currentY = curNumCoordList[j]
+                nextX, nextY = curNumCoordList[j + 1]
+                #indexes are in order, if we jump more than 1 square then there must be a disconnect or doubleback
+                if (currentX + 1 == nextX and currentY == nextY) or \
+                   (currentY + 1 == nextY and currentX == nextX) or \
+                   (currentX - 1 == nextX and currentY == nextY) or \
+                   (currentY - 1 == nextY and currentX == nextX):
+                    continue
+                else:#number must not be connected
+                    isValid = False
+                    break
+
+            if (isValid):
+                connectedNumbers+=1
+
+        return connectedNumbers
+
     #placeholder
     def DetermineFitness(self, individual):
-        return 0
+        #Fitness will be maximized, the maximal value will be according to the following formula
+        #fitMax = #ofNumsToConnect * weight + totalSquares
+        #The formula to calculate fitness will be as follows:
+        #fit = fitMax - #ofUnconnectedPairs * weight - #ofEmptySquares
+        weightOfConnectedness = 10
```

Comment:
This is a ratio. Currently, a connected set of numbers is 10* more important the empty square. May need to be higher, as an empty square is pretty easy to fill. May want to even consider ignoring empty squares until the end. If we can connect all cities, I think filling in blanks should be simple